hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSDirRenameOp.java

```diff
@@ -574,7 +580,28 @@ private static void validateRenameSource(FSDirectory fsd,
     }
     // srcInode and its subtree cannot contain snapshottable directories with
     // snapshots
-    FSDirSnapshotOp.checkSnapshot(fsd, srcIIP, null);
+    FSDirSnapshotOp.checkSnapshot(fsd, srcIIP, snapshottableDirs);
+  }
+
+  private static void validateNestSnapshot(FSDirectory fsd, INodeDirectory dstParent,
+      List<INodeDirectory> snapshottableDirs) throws SnapshotException {
+
+    if (fsd.getFSNamesystem().getSnapshotManager().isAllowNestedSnapshots()) {
+      return;
+    }
+
+    /*
+     * snapshottableDirs is a list of snapshottable directory(child of rename src)
+     * which do not have snapshots yet. If this list is not empty, that means rename
+     * src have snapshottable descendant directories.
```

Comment:
if the list is not empty, it implies the rename src have snapshottable descendant directories that do not have snapshots. What if there are descendant dirs with snapshots? Would it be allows to rename?

Reply:
If any of the descendants dirs got some snapshot, `FSDirRenameOp#validateRenameSource` will finally call `FSDirSnapshotOp#checkSnapshot` and throw `SnapshotException`. And this check happens before this one.